### PR TITLE
Stop using VersionedObject in resource.Builder, use unstructured

### PR DESCRIPTION
**Pull request by smarterclayton**
Remove the need for kubectl callers to distinguish between unstructured and versioned / type aware builders.  The factory should create a single builder than can be set to return unstructured objects.  Callers can then use one of the new helpers on `resource.Info` to convert the objects into the desired form - `Internal()` for printers, `Typed()` for external versions, and `Unstructured()` to ensure the object is in the right state.  Leverages the new scheme support for unstructured conversion so that higher level callers can perform best effort conversion (get typed versions if you have them, otherwise use default behavior).  

`get.go` demonstrates this by removing the previous logic that depended on the underlying scheme.  Other commands are updated to be consistent.

Includes #55650 and #55647.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@kubernetes/sig-cli-api-reviews

File: pkg/kubectl/cmd/set/set_env.go [hunk 1]
```diff
@@ -309,21 +309,22 @@ func (o *EnvOptions) RunEnv(f cmdutil.Factory) error {
...
+		b = b.Local()
 	}
 
 	o.Infos, err = b.Do().Infos()
 	if err != nil {
 		return err
 	}
 	patches := CalculatePatches(o.Infos, o.Encoder, func(info *resource.Info) ([]byte, error) {
-		_, err := o.UpdatePodSpecForObject(info.VersionedObject, func(spec *v1.PodSpec) error {
+		info.Object = info.AsVersioned()
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_env.go]: this will ignore convert err. when convert error will return `the object is not a pod or does not have a pod template` for users. i think this error cannot help users...

but i am ok use `info.AsVersioned()` in `PrintObject`

Comment (smarterclayton) [reply pkg/kubectl/cmd/set/set_env.go]: A convert error can only happen if the object is not a valid type, which means we aren't compiled with it, which means we weren't going to be able to support it here anyway.  We already have types that aren't compiled in, so the helper already has to deal with that.

Comment (smarterclayton) [reply pkg/kubectl/cmd/set/set_env.go]: I would say that we should be as explicit as we can when something is not in the known type.  We should probably do that up front in the resource.Builder (have a RequireVersioned() flag) that will accumulate errors for anything that isn't versioned.

However, when doing anything from disk we have to be pretty tolerate of things that aren't versioned (i.e. `kubectl set image -f FILE_THAT_HAS_MIXED_OBJECTS`) - that command shouldn't fail for objects it doesn't recognize.  So tolerating untyped objects in set commands is going to be pretty important.

**Comment by smarterclayton** [comment]
@deads2k UnstructuredObject() is now dead.  Can you look at the last two commits and give your blessing / opinion?

**Comment by smarterclayton** [comment]
Last commit added that forces all resource.Builder users to decide on Unstructured or Internal.  I'll resist the urge to keep refactoring at this point - this is enough to shrink get.go, give us a clear anchor point to remove unstructured, and removes the accrued debt around not being able to rationalize unstructured and versioned types.

File: pkg/kubectl/resource/builder.go [hunk 2]
```diff
@@ -166,16 +171,42 @@ func (b *Builder) FilenameParam(enforceNamespace bool, filenameOptions *Filename
...
 	}
 	b.mapper = b.unstructured
 	return b
 }
 
+// Internal updates the builder so that it will convert objects off the wire
+// into the internal form if necessary. Using internal types is lossy - fields added
+// to the server will not be seen by the client code and may result in failure. Only
+// use this mode when working offline, or when generating patches to send to the server.
+// TODO: Introduce a Versioned() form that performs this same function, but biases
```
Comment (deads2k) [inline pkg/kubectl/resource/builder.go]: The warning on this method would remain identical.  For a given snapshot of external types, the internal ones are just as good.

File: pkg/kubectl/cmd/annotate.go [hunk 3]
```diff
@@ -189,13 +189,10 @@ func (o AnnotateOptions) RunAnnotate(f cmdutil.Factory, cmd *cobra.Command) erro
...
 	includeUninitialized := cmdutil.ShouldIncludeUninitialized(cmd, false)
-	b := f.NewBuilder()
-	if o.local {
-		b = b.Local()
-	} else {
-		b = b.Unstructured()
-	}
-	b = b.ContinueOnError().
+	b := f.NewBuilder().
+		Unstructured().
```
Comment (deads2k) [inline pkg/kubectl/cmd/annotate.go]: @juanvallejo  didn't `Unstructured`  attempt to make a remote call and that broke `--local` since it tried to contact the server.

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 4]
```diff
@@ -112,7 +112,7 @@ func (o *SetLastAppliedOptions) Complete(f cmdutil.Factory, cmd *cobra.Command)
 	o.Codec = f.JSONEncoder()
 
 	var err error
-	o.Mapper, o.Typer = f.UnstructuredObject()
+	o.Mapper, o.Typer = f.Object()
```
Comment (deads2k) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: I don't have all the code in my head, but this leaps out as moving in the wrong direction.  Why the change?

File: pkg/kubectl/cmd/annotate.go (hunk 3 above)
Comment (juanvallejo) [reply pkg/kubectl/cmd/annotate.go]: Yes, I believe `Unstructured()` makes use of the discovery client to retrieve mappings

File: pkg/kubectl/cmd/set/set_serviceaccount.go [hunk 5]
```diff
@@ -127,16 +127,16 @@ func (saConfig *serviceAccountConfig) Complete(f cmdutil.Factory, cmd *cobra.Com
 	saConfig.serviceAccountName = args[len(args)-1]
 	resources := args[:len(args)-1]
 	includeUninitialized := cmdutil.ShouldIncludeUninitialized(cmd, false)
-	builder := f.NewBuilder().ContinueOnError().
+	builder := f.NewBuilder().
+		LocalParam(saConfig.local).
+		ContinueOnError().
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_serviceaccount.go]: intentional addition?  I like it, but wanted to make sure it was on purpose.

File: pkg/kubectl/cmd/util/factory_builder.go [hunk 6]
```diff
@@ -108,12 +103,6 @@ func (f *ring2Factory) PrintObject(cmd *cobra.Command, isLocal bool, mapper meta
 	_, typer := f.objectMappingFactory.Object()
 	gvks, _, err := typer.ObjectKinds(obj)
 
-	// fall back to an unstructured object if we get something unregistered
-	if runtime.IsNotRegisteredError(err) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_builder.go]: I remember this being important for handling cases where you querying aggregated types and CRDs.

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 7]
```diff
@@ -73,33 +73,15 @@ func NewObjectMappingFactory(clientAccessFactory ClientAccessFactory) ObjectMapp
...
-
-// TODO: unstructured and structured discovery should both gracefully degrade in the presence
-// of errors, and it should be possible to get access to the unstructured object typers no matter
-// whether the server responds or not. Make the legacy scheme recognize object typer, then we can
-// safely fall back to the built in restmapper as a last resort.
-func (f *ring1Factory) unstructuredObjectLoader() (meta.RESTMapper, runtime.ObjectTyper, error) {
-	discoveryClient, err := f.clientAccessFactory.DiscoveryClient()
-	if err != nil {
-		return nil, nil, err
+		glog.V(3).Infof("Unable to get a discovery client to find server resources, falling back to hardcoded types: %v", err)
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Why should we continue doing this?

File: pkg/kubectl/cmd/annotate.go (hunk 3 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/annotate.go]: Yes, and now unstructured fails back to using the innate discovery.

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 4 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/apply_set_last_applied.go]: There is only Object() now.

File: pkg/kubectl/cmd/util/discovery.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+	err    error
+	mapper meta.RESTMapper
+	typer  runtime.ObjectTyper
+}
+
+// NewLazyObjectLoader handles unrecoverable errors when creating a RESTMapper / ObjectTyper by
+// returning those initialization errors when the interface methods are invoked. This defers the
+// initialization and any server calls until a client actually needs to perform the action.
+func NewLazyObjectLoader(fn func() (meta.RESTMapper, runtime.ObjectTyper, error)) (meta.RESTMapper, runtime.ObjectTyper) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/discovery.go]: Ah, I see.  This is what is letting you remove the special local blocks.

File: pkg/kubectl/cmd/apply_set_last_applied.go (hunk 4 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/apply_set_last_applied.go]: Everything is in the same path.  You ask the builder whether you want Unstructured() or Internal().  You can't opt out.

File: pkg/kubectl/cmd/set/set_serviceaccount.go (hunk 5 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/set/set_serviceaccount.go]: Yes, was down below.

File: pkg/kubectl/cmd/util/discovery.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	obj := &lazyObject{loader: fn}
+	return obj, obj
+}
+
+// init lazily loads the mapper and typer, returning an error if initialization has failed.
+func (o *lazyObject) init() error {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.err != nil || o.loader == nil {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/discovery.go]: separate the if blocks and set the error if the loader is nil.

File: pkg/kubectl/cmd/set/set_serviceaccount.go (hunk 5 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/set/set_serviceaccount.go]: I have this TODO somewhere, but local shouldn't be a command specific flag.  It should be a global and enforced at multiple levels.  There are decisions that Factory methods make that should depend on local.

File: pkg/kubectl/cmd/util/factory_builder.go (hunk 6 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/util/factory_builder.go]: I now handle unknown types in the fundamental scheme safely for ObjectKinds.

File: pkg/kubectl/cmd/util/discovery.go (hunk 9 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/util/discovery.go]: No, loader in this case is the indicator that we have already loaded or not.  We clear the loader and then don't continue.

File: pkg/kubectl/cmd/util/discovery.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+		return o.err
+	}
+	o.mapper, o.typer, o.err = o.loader()
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/discovery.go]: check the mapper and typer early and return if they're set.

File: pkg/kubectl/cmd/util/discovery.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+	o.loader = nil
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/discovery.go]: I don't see a need to do this.  mapper and typer shouldn't be nil out of the loader.

File: pkg/kubectl/cmd/util/factory_object_mapping.go (hunk 7 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/util/factory_object_mapping.go]: Fallback for local.  You still need a RESTMapper if you can't contact the server.  If discovery doesn't work, all but a few commands want to fail partially.

File: pkg/kubectl/cmd/util/discovery.go (hunk 11 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/util/discovery.go]: This is the indicator that covers whether loading has happened.

File: pkg/kubectl/cmd/util/discovery.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+	return o.err
+}
+
+var _ meta.RESTMapper = &lazyObject{}
+var _ runtime.ObjectTyper = &lazyObject{}
+
+func (o *lazyObject) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/discovery.go]: I'd like to the deferred RESTMapper moved with the other RESTMappers and not specific to kubectl.

File: pkg/kubectl/cmd/util/factory_builder.go [hunk 13]
```diff
@@ -163,7 +155,7 @@ func (f *ring2Factory) NewBuilder() *resource.Builder {
 	mapper, typer := f.objectMappingFactory.Object()
 
 	unstructuredClientMapperFunc := resource.ClientMapperFunc(f.objectMappingFactory.UnstructuredClientForMapping)
-	unstructuredMapper, unstructuredTyper, _ := f.objectMappingFactory.UnstructuredObject()
+	unstructuredMapper, unstructuredTyper := f.objectMappingFactory.Object()
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_builder.go]: this looks backwards

File: pkg/kubectl/cmd/util/discovery.go (hunk 12 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/util/discovery.go]: it's gotta have object typer too, but I can do that.

File: pkg/kubectl/cmd/util/discovery.go (hunk 11 above)
Comment (deads2k) [reply pkg/kubectl/cmd/util/discovery.go]: > This is the indicator that covers whether loading has happened.

Using it as the trigger opens you up to accidentally leaving nil.  Seems like triggering the other way around avoids that problem.

**Comment by mengqiy** [comment]
bazel test is failing.

File: pkg/kubectl/cmd/util/factory_builder.go [hunk 14]
```diff
@@ -51,17 +50,9 @@ func NewBuilderFactory(clientAccessFactory ClientAccessFactory, objectMappingFac
 func (f *ring2Factory) PrinterForCommand(cmd *cobra.Command, isLocal bool, outputOpts *printers.OutputOptions, options printers.PrintOptions) (printers.ResourcePrinter, error) {
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/factory_builder.go]: > isLocal bool

It seems we don't use it any more. We can drop it.

File: pkg/kubectl/resource/visitor.go [hunk 15]
```diff
@@ -81,10 +82,6 @@ type Info struct {
 	// Optional, Source is the filename or URL to template file (.json or .yaml),
 	// or stdin to use to handle the resource
 	Source string
-	// Optional, this is the provided object in a versioned type before defaulting
-	// and conversions into its corresponding internal type. This is useful for
-	// reflecting on user intent which may be lost after defaulting and conversions.
-	VersionedObject runtime.Object
 	// Optional, this is the most recent value returned by the server if available
```
Comment (mengqiy) [inline pkg/kubectl/resource/visitor.go]: Can you please update the comment to make it more helpful?

**Review by mengqiy — COMMENTED** [review]
I like this PR!

File: pkg/kubectl/cmd/util/factory_builder.go (hunk 14 above)
Comment (smarterclayton) [reply pkg/kubectl/cmd/util/factory_builder.go]: Will do that in a follow up - was going to do the following:

1. remove isLocal
2. Collapse OutputOptions into PrintOptions
3. show how that is used by server-side get rendering

File: pkg/kubectl/resource/visitor.go (hunk 15 above)
Comment (smarterclayton) [reply pkg/kubectl/resource/visitor.go]: Yes

File: pkg/kubectl/cmd/util/factory_builder.go (hunk 14 above)
Comment (juanvallejo) [reply pkg/kubectl/cmd/util/factory_builder.go]: > 2. Collapse OutputOptions into PrintOptions

PR currently open for this: https://github.com/kubernetes/kubernetes/pull/54811

**Comment by smarterclayton** [comment]
Updated and rebased on the latest changes.  Last thing blocking client side get alpha changes for 1.9

File: pkg/kubectl/resource/mapper.go [hunk 16]
```diff
@@ -44,17 +34,27 @@ type Mapper struct {
...
+	copied.RESTMapper = NewRelaxedRESTMapper(m.RESTMapper)
+	copied.ClientMapper = NewRelaxedClientMapper(m.ClientMapper)
+	return &copied
+}
+
 // InfoForData creates an Info object for the given data. An error is returned
 // if any of the decoding or client lookup steps fail. Name and namespace will be
 // set into Info if the mapping's MetadataAccessor can retrieve them.
 func (m *Mapper) InfoForData(data []byte, source string) (*Info, error) {
-	versions := &runtime.VersionedObjects{}
```
Comment (shiywang) [inline pkg/kubectl/resource/mapper.go]: Is kubectl the only consumer of `VersionedObjects{}` do we also need to remove [here](https://github.com/kubernetes/kubernetes/blame/3df3c580b7a318c8fd166296b61d891f6c0359fa/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L93) and append logic [there](https://github.com/kubernetes/kubernetes/blame/3df3c580b7a318c8fd166296b61d891f6c0359fa/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L135-L138) ?

Comment (shiywang) [reply pkg/kubectl/resource/mapper.go]: Sorry, i think i forget the compatible problem...please omit

Comment (smarterclayton) [reply pkg/kubectl/resource/mapper.go]: I will be removing VersionedObjects in a future PR.
